[Flaky Test] SnapshotResiliencyTests.testConcurrentSnapshotDeleteAndDeleteIndex if flaky

https://build.ci.opensearch.org/job/gradle-check/20332/

```
REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.snapshots.SnapshotResiliencyTests.testConcurrentSnapshotDeleteAndDeleteIndex" -Dtests.seed=B7F15DFF786CE73D -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es-NI -Dtests.timezone=Asia/Oral -Druntime.java=17
NOTE: leaving temporary files on disk at: /var/jenkins/workspace/gradle-check/search/server/build/testrun/test/temp/org.opensearch.snapshots.SnapshotResiliencyTests_B7F15DFF786CE73D-001
NOTE: test params are: codec=Asserting(Lucene95): {}, docValues:{}, maxPointsInLeafNode=330, maxMBSortInHeap=6.1702426242178525, sim=Asserting(RandomSimilarity(queryNorm=true): {}), locale=es-NI, timezone=Asia/Oral
NOTE: Linux 5.15.0-1036-aws amd64/Eclipse Adoptium 17.0.7 (64-bit)/cpus=96,threads=1,free=189188400,total=536870912
```

https://github.com/opensearch-project/OpenSearch/pull/12335#issuecomment-1946461615:

```
REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.snapshots.SnapshotResiliencyTests.testConcurrentSnapshotDeleteAndDeleteIndex" -Dtests.seed=E816C7242330BF50 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=ms-MY -Dtests.timezone=Europe/Athens -Druntime.java=21

org.opensearch.snapshots.SnapshotResiliencyTests > testConcurrentSnapshotDeleteAndDeleteIndex FAILED
    java.lang.AssertionError: java.lang.StringIndexOutOfBoundsException: Range [13, 2) out of bounds for length 6
        at __randomizedtesting.SeedInfo.seed([E816C7242330BF50:BD5D3283E0ED3D1D]:0)
        at org.opensearch.repositories.blobstore.BlobStoreRepository.lambda$executeOneStaleIndexDelete$40(BlobStoreRepository.java:1709)
        at org.opensearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:74)
        at org.opensearch.action.ActionRunnable$2.doRun(ActionRunnable.java:89)
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
        at org.opensearch.cluster.coordination.AbstractCoordinatorTestCase$1.run(AbstractCoordinatorTestCase.java:1498)
        at org.opensearch.cluster.coordination.DeterministicTaskQueue.runTask(DeterministicTaskQueue.java:147)
        at org.opensearch.cluster.coordination.DeterministicTaskQueue.runRandomTask(DeterministicTaskQueue.java:141)
        at org.opensearch.cluster.coordination.DeterministicTaskQueue.runAllRunnableTasks(DeterministicTaskQueue.java:91)
        at org.opensearch.snapshots.SnapshotResiliencyTests.runUntil(SnapshotResiliencyTests.java:1613)
        at org.opensearch.snapshots.SnapshotResiliencyTests.verifyReposThenStopServices(SnapshotResiliencyTests.java:335)

        Caused by:
        java.lang.StringIndexOutOfBoundsException: Range [13, 2) out of bounds for length 6
            at java.****/jdk.internal.util.Preconditions$1.apply(Preconditions.java:55)
            at java.****/jdk.internal.util.Preconditions$1.apply(Preconditions.java:52)
            at java.****/jdk.internal.util.Preconditions$4.apply(Preconditions.java:213)
            at java.****/jdk.internal.util.Preconditions$4.apply(Preconditions.java:210)
            at java.****/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:98)
            at java.****/jdk.internal.util.Preconditions.outOfBoundsCheckFromToIndex(Preconditions.java:112)
            at java.****/jdk.internal.util.Preconditions.checkFromToIndex(Preconditions.java:349)
            at java.****/java.lang.String.checkBoundsBeginEnd(String.java:4865)
            at java.****/java.lang.String.substring(String.java:2834)
            at org.opensearch.repositories.blobstore.BlobStoreRepository.lambda$executeOneStaleIndexDelete$40(BlobStoreRepository.java:1658)
            ... 9 more
```

--- a/server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java
+++ b/server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java
@@ -1592,8 +1592,11 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
                                 Map<String, BlobMetadata> shardLevelBlobs = shardBlob.getValue().listBlobs();
                                 for (Map.Entry<String, BlobMetadata> shardLevelBlob : shardLevelBlobs.entrySet()) {
                                     String blob = shardLevelBlob.getKey();
-                                    String snapshotUUID = blob.substring(SHALLOW_SNAPSHOT_PREFIX.length(), blob.length() - ".dat".length());
                                     if (blob.startsWith(SHALLOW_SNAPSHOT_PREFIX) && blob.endsWith(".dat")) {
+                                        String snapshotUUID = blob.substring(
+                                            SHALLOW_SNAPSHOT_PREFIX.length(),
+                                            blob.length() - ".dat".length()
+                                        );
                                         RemoteStoreShardShallowCopySnapshot remoteStoreShardShallowCopySnapshot =
                                             REMOTE_STORE_SHARD_SHALLOW_COPY_SNAPSHOT_FORMAT.read(
                                                 shardBlob.getValue(),

--- a/server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java
+++ b/server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java
@@ -1116,10 +1116,7 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
                                 String indexId = fileToDeletePath[1];
                                 String shardId = fileToDeletePath[2];
                                 String shallowSnapBlob = fileToDeletePath[3];
-                                String snapshotUUID = shallowSnapBlob.substring(
-                                    SHALLOW_SNAPSHOT_PREFIX.length(),
-                                    shallowSnapBlob.length() - ".dat".length()
-                                );
+                                String snapshotUUID = extractShallowSnapshotUUID(shallowSnapBlob).orElseThrow();
                                 BlobContainer shardContainer = blobStore().blobContainer(indicesPath().add(indexId).add(shardId));
                                 RemoteStoreShardShallowCopySnapshot remoteStoreShardShallowCopySnapshot =
                                     REMOTE_STORE_SHARD_SHALLOW_COPY_SNAPSHOT_FORMAT.read(
@@ -1586,47 +1583,43 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
                 try {
                     logger.debug("[{}] Found stale index [{}]. Cleaning it up", metadata.name(), indexSnId);
                     if (remoteStoreLockManagerFactory != null) {
-                        Map<String, BlobContainer> shardBlobs = indexEntry.getValue().children();
-                        if (!shardBlobs.isEmpty()) {
-                            for (Map.Entry<String, BlobContainer> shardBlob : shardBlobs.entrySet()) {
-                                Map<String, BlobMetadata> shardLevelBlobs = shardBlob.getValue().listBlobs();
-                                for (Map.Entry<String, BlobMetadata> shardLevelBlob : shardLevelBlobs.entrySet()) {
-                                    String blob = shardLevelBlob.getKey();
-                                    if (blob.startsWith(SHALLOW_SNAPSHOT_PREFIX) && blob.endsWith(".dat")) {
-                                        String snapshotUUID = blob.substring(
-                                            SHALLOW_SNAPSHOT_PREFIX.length(),
-                                            blob.length() - ".dat".length()
+                        final Map<String, BlobContainer> shardBlobs = indexEntry.getValue().children();
+                        for (Map.Entry<String, BlobContainer> shardBlob : shardBlobs.entrySet()) {
+                            for (String blob : shardBlob.getValue().listBlobs().keySet()) {
+                                final Optional<String> snapshotUUID = extractShallowSnapshotUUID(blob);
+                                if (snapshotUUID.isPresent()) {
+                                    RemoteStoreShardShallowCopySnapshot remoteStoreShardShallowCopySnapshot =
+                                        REMOTE_STORE_SHARD_SHALLOW_COPY_SNAPSHOT_FORMAT.read(
+                                            shardBlob.getValue(),
+                                            snapshotUUID.get(),
+                                            namedXContentRegistry
                                         );
-                                        RemoteStoreShardShallowCopySnapshot remoteStoreShardShallowCopySnapshot =
-                                            REMOTE_STORE_SHARD_SHALLOW_COPY_SNAPSHOT_FORMAT.read(
-                                                shardBlob.getValue(),
-                                                snapshotUUID,
-                                                namedXContentRegistry
-                                            );
-                                        String indexUUID = remoteStoreShardShallowCopySnapshot.getIndexUUID();
-                                        String remoteStoreRepoForIndex = remoteStoreShardShallowCopySnapshot.getRemoteStoreRepository();
-                                        // Releasing lock files before deleting the shallow-snap-UUID file because in case of any failure
-                                        // while releasing the lock file, we would still have the corresponding shallow-snap-UUID file
-                                        // and that would be used during next delete operation for releasing this stale lock file
-                                        RemoteStoreLockManager remoteStoreMetadataLockManager = remoteStoreLockManagerFactory
-                                            .newLockManager(remoteStoreRepoForIndex, indexUUID, shardBlob.getKey());
-                                        remoteStoreMetadataLockManager.release(
-                                            FileLockInfo.getLockInfoBuilder().withAcquirerId(snapshotUUID).build()
-                                        );
-                                        if (!isIndexPresent(clusterService, indexUUID)) {
-                                            // this is a temporary solution where snapshot deletion triggers remote store side
-                                            // cleanup if index is already deleted. We will add a poller in future to take
-                                            // care of remote store side cleanup.
-                                            // see https://github.com/opensearch-project/OpenSearch/issues/8469
-                                            new RemoteSegmentStoreDirectoryFactory(
-                                                remoteStoreLockManagerFactory.getRepositoriesService(),
-                                                threadPool
-                                            ).newDirectory(
-                                                remoteStoreRepoForIndex,
-                                                indexUUID,
-                                                new ShardId(Index.UNKNOWN_INDEX_NAME, indexUUID, Integer.valueOf(shardBlob.getKey()))
-                                            ).close();
-                                        }
+                                    String indexUUID = remoteStoreShardShallowCopySnapshot.getIndexUUID();
+                                    String remoteStoreRepoForIndex = remoteStoreShardShallowCopySnapshot.getRemoteStoreRepository();
+                                    // Releasing lock files before deleting the shallow-snap-UUID file because in case of any failure
+                                    // while releasing the lock file, we would still have the corresponding shallow-snap-UUID file
+                                    // and that would be used during next delete operation for releasing this stale lock file
+                                    RemoteStoreLockManager remoteStoreMetadataLockManager = remoteStoreLockManagerFactory.newLockManager(
+                                        remoteStoreRepoForIndex,
+                                        indexUUID,
+                                        shardBlob.getKey()
+                                    );
+                                    remoteStoreMetadataLockManager.release(
+                                        FileLockInfo.getLockInfoBuilder().withAcquirerId(snapshotUUID.get()).build()
+                                    );
+                                    if (!isIndexPresent(clusterService, indexUUID)) {
+                                        // this is a temporary solution where snapshot deletion triggers remote store side
+                                        // cleanup if index is already deleted. We will add a poller in future to take
+                                        // care of remote store side cleanup.
+                                        // see https://github.com/opensearch-project/OpenSearch/issues/8469
+                                        new RemoteSegmentStoreDirectoryFactory(
+                                            remoteStoreLockManagerFactory.getRepositoriesService(),
+                                            threadPool
+                                        ).newDirectory(
+                                            remoteStoreRepoForIndex,
+                                            indexUUID,
+                                            new ShardId(Index.UNKNOWN_INDEX_NAME, indexUUID, Integer.parseInt(shardBlob.getKey()))
+                                        ).close();
                                     }
                                 }
                             }
@@ -3365,12 +3358,7 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
                             blob.substring(SNAPSHOT_PREFIX.length(), blob.length() - ".dat".length())
                         ) == false)
                     || (remoteStoreLockManagerFactory != null
-                        ? (blob.startsWith(SHALLOW_SNAPSHOT_PREFIX)
-                            && blob.endsWith(".dat")
-                            && survivingSnapshotUUIDs.contains(
-                                blob.substring(SHALLOW_SNAPSHOT_PREFIX.length(), blob.length() - ".dat".length())
-                            ) == false)
-                        : false)
+                        && extractShallowSnapshotUUID(blob).map(survivingSnapshotUUIDs::contains).orElse(false))
                     || (blob.startsWith(UPLOADED_DATA_BLOB_PREFIX) && updatedSnapshots.findNameFile(canonicalName(blob)) == null)
                     || FsBlobContainer.isTempBlobName(blob)
             )
@@ -3510,6 +3498,13 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
                 logger.warn("store cannot be marked as corrupted", inner);
             }
         }
+    }
+
+    private static Optional<String> extractShallowSnapshotUUID(String blobName) {
+        if (blobName.startsWith(SHALLOW_SNAPSHOT_PREFIX)) {
+            return Optional.of(blobName.substring(SHALLOW_SNAPSHOT_PREFIX.length(), blobName.length() - ".dat".length()));
+        }
+        return Optional.empty();
     }
 
     /**

